### PR TITLE
support exists/in subquery convert to semi-join (#864)

### LINKED ISSUE #864: feature:  support semi-join in stonedb, include EXISTS and IN subqueries
**Is your feature request related to a problem? Please describe.**

The reason why our tpch Q4 query is slow is because the execution strategy of subquery is only one kind of nested loop, here we want to do an optimization is to convert the subquery expansion to semi-join, can use hash join, so that we can use hash-join to speed up.
**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by duanfuxiang0**
## Summary about this PR

Issue Number: close #864 

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [x] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Review by isredstar — APPROVED** [review]

**Review by hustjieke — CHANGES_REQUESTED** [review]
Please do a rebase, the same commit msg is too much

**Review by konghaiya — APPROVED** [review]
LGTM

File: mysql-test/suite/tianmu/r/issue947.result [hunk 1]
```diff
@@ -34,15 +34,11 @@ a
 SELECT * FROM test1 WHERE a IN ( SELECT b FROM test2 WHERE b=1);
 a
 1
-1
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/issue947.result:37]: @davidshiz , pls check out this MTRs. Make sure why we change the test result.

File: mysql-test/suite/tianmu/r/semi-join.result [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+6
+7
+SELECT t1_int FROM t1 WHERE EXISTS (SELECT t2_int FROM t2 WHERE t1.t1_int = t2.t2_int) ORDER BY t1_int;
+t1_int
+1
+3
+5
+6
+7
+SELECT t1_int, count(*) FROM t1 WHERE EXISTS (SELECT t2_int FROM t2 WHERE t1.t1_int = t2.t2_int) GROUP BY t1_int;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/semi-join.result:62]: Add you add tests to display the query plan.  therefore, we can identify whether semi-join takes into effect

**Comment by RingsC** [comment]
> ## Summary about this PR
> Issue Number: close #864
> 
> ## Tests Check List
> * [ ]  Unit test
> * [ ]  Integration test
> * [x]  Manual test (add detailed scripts or steps below)
> * [ ]  No code
> 
> ## Changelog
> * [ ]  New Feature
> * [ ]  Bug Fix
> * [x]  Performance Improvement
> * [ ]  Build/Testing/CI/CD
> * [ ]  Documentation
> * [ ]  Not for changelog (changelog entry is not required)
> 
> ## Documentation
> * [ ]  Affects user behaviors
> * [ ]  Contains syntax changes
> * [ ]  Contains variable changes
> * [ ]  Contains experimental features

@duanfuxiang0 pls, give more summary info on this PR.

File: mysql-test/suite/tianmu/r/semi-join.result [hunk 3]
```diff
@@ -0,0 +1,69 @@
+DROP DATABASE IF EXISTS test_db1;
+CREATE DATABASE test_db1;
+USE test_db1;
+CREATE TABLE t1 (t1_int INT, t1_char CHAR(5), PRIMARY KEY (t1_int)) ENGINE=TIANMU;
+CREATE TABLE t2 (t2_int INT, t2_char CHAR(5)) ENGINE=TIANMU;
+INSERT INTO t1 VALUES (1, 'aaa'),(2, 'aaa'),(3, 'ccc'),(4, 'ddd'),(5, 'aaa'),(6, ''),(7, 'eee');
+INSERT INTO t2 VALUES (1, 'eee'),(3, 'ccc'),(5, 'jjj'),(6, ''),(7, 'lll'),(9, 'eee'),(11, 'nnn');
+SELECT t1_int FROM t1 WHERE t1_int IN (SELECT t2_int FROM t2 WHERE t1.t1_int = t2.t2_int);
+t1_int
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/semi-join.result:9]: @davidshiz ,pls give us more complex query tests for `in/exists` statement. Thanks

**Review by RingsC — CHANGES_REQUESTED** [review]

File: mysql-test/suite/tianmu/r/issue947.result (hunk 1 above)
Comment (davidshiz) [reply mysql-test/suite/tianmu/r/issue947.result:37]: ok

Comment (davidshiz) [reply mysql-test/suite/tianmu/r/issue947.result:37]: after verification, the issue947.result was wrong before, it is correct now @RingsC @duanfuxiang0

**Review by wisehead — APPROVED** [review]

**Pull request closed without merging**